### PR TITLE
Proxy Celostats fixes

### PR DESCRIPTION
**Pull request by gastonponti**
### Description

The proxy sends stats to celostats, previously signed by the validator

[WIP] Proxies stop sending data after some time

### Other changes

The validator node it doesn't know anymore which proxy node is the one that sends stats
(This allow us to remove a stats-proxy, and add another one with that configuration)

Adds a refactor done in the PR #1033 

### Tested

Manually

### Related issues

- Part of #1026 

### Backwards compatibility

Against the Multiproxy branch, it removes a flag.

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -780,11 +780,6 @@ var (
 		Usage: "Each enode URL in a pair is separated by a semicolon. Enode URL pairs are separated by a space. The format should be \"<proxy 0 internal facing enode URL>;<proxy 0 external facing enode URL>,<proxy 1 internal facing enode URL>;<proxy 1 external facing enode URL>,...\" (deprecated, use --proxy.proxyenodeurlpairs)",
 	}
 
-	ProxyCeloStatsFlag = cli.StringFlag{
```
Comment (trianglesphere) [inline cmd/utils/flags.go:783]: We should probably note that this flag is deprecated and doesn't do anything rather than immediately removing it as passing in undefined flags can cause problems.

**Review by trianglesphere — COMMENTED** [review]
Looks good so far. One comment on flags. Also gofmt everything.

File: ethstats/ethstats.go [hunk 2]
```diff
@@ -96,52 +97,68 @@ type blockChain interface {
...
 	if ethServ != nil {
 		engine = ethServ.Engine()
 	} else {
 		engine = lesServ.Engine()
 	}
 
 	backend := engine.(*istanbulBackend.Backend)
+	if !backend.IsProxiedValidator() {
+		// Parse the netstats connection url
+		re := regexp.MustCompile("([^:@]*)?@(.+)")
```
Comment (mcortesi) [inline ethstats/ethstats.go:144]: good opportunity to extract function refactor :)

File: ethstats/ethstats.go [hunk 3]
```diff
@@ -546,7 +571,10 @@ func (s *Service) readLoop(conn *websocket.Conn) {
 			log.Info("Ping from websocket server", "msg", msg)
 			// Primus server might want to have a pong or it closes the connection
 			var serverTime = fmt.Sprintf("primus::pong::%d", time.Now().UnixNano()/int64(time.Millisecond))
-			conn.WriteJSON(serverTime)
+			err := conn.WriteMessage(websocket.TextMessage, []byte(serverTime))
```
Comment (mcortesi) [inline ethstats/ethstats.go:574]: why this change?

File: ethstats/ethstats.go [hunk 4]
```diff
@@ -1035,8 +1061,8 @@ func (s *Service) reportStats(conn *websocket.Conn) error {
 	log.Trace("Sending node details to ethstats")
 
 	stats := map[string]interface{}{
-		"id":      s.node,
-		"address": etherBase,
+		"id":      s.backend.ValidatorAddress().String(),
+		"address": validatorAddress,
```
Comment (mcortesi) [inline ethstats/ethstats.go:1065]: if eth == nil validatorAddress is empty. Can s.eth be nil? shouldn't that imply we need to skip or something?

File: ethstats/ethstats.go [hunk 5]
```diff
@@ -1035,8 +1061,8 @@ func (s *Service) reportStats(conn *websocket.Conn) error {
 	log.Trace("Sending node details to ethstats")
 
 	stats := map[string]interface{}{
-		"id":      s.node,
-		"address": etherBase,
+		"id":      s.backend.ValidatorAddress().String(),
```
Comment (mcortesi) [inline ethstats/ethstats.go:1064]: if last comment is solved, this could be `validatorAddress.String()`

File: ethstats/ethstats.go [hunk 6]
```diff
@@ -444,7 +456,20 @@ func (s *Service) login(conn *websocket.Conn, sendCh chan *StatsPayload) error {
 	return nil
 }
 
-func (s *Service) handleDelegateSign(messageToSign *StatsPayload) error {
+func (s *Service) waitAndDelegateMessageWithTimeout(conn *websocket.Conn, sendCh chan *StatsPayload) error {
+	select {
+	case signedMessage := <-sendCh:
+		err := s.handleDelegateSend(conn, signedMessage)
```
Comment (mcortesi) [inline ethstats/ethstats.go:462]: can replace 
```go
		err := s.handleDelegateSend(conn, signedMessage)
		if err != nil {
			return err
    }
```
with:
```go
		return s.handleDelegateSend(conn, signedMessage)
```
and remove the `return nil` at the end

File: ethstats/ethstats.go [hunk 7]
```diff
@@ -238,26 +249,33 @@ func (s *Service) loop() {
 			case <-headSub.Err():
 				break HandleLoop
 			case delegateSignMsg := <-istDelegateSignCh:
```
Comment (mcortesi) [inline ethstats/ethstats.go:251]: one nice (and small) refactor is to treat istDelegateCh as a "router". IMO on line 206 where we subscribe, we could just launch a go routiner like:
```go
go func() {
  ch := make(chan istanbul.MessageWithPeerIDEvent, istDelegateSignChanSize)
	subscription := s.backend.SubscribeNewDelegateSignEvent(ch)
  defer subscription.Unsubscribe()
  
  for {
    case msg := <- ch:
      // route msg either to signCh or sendCh
    case err := <- subscription.Err() // Handle subscrition errors that we are currently not doing!
    case <- quit // need to verify this, but some signal to stop (using go context module can work too)
  }
}
```

**Comment by oneeman** [comment]
With these changes, I think the following fields (and all references to them) can be removed:
* `ProxyConfig.StatsHandler`
* `Proxy.statsHandler`
* `ProxyInfo.StatsHandler`

I believe the only one of them that's exposed externally somewhat is `ProxyInfo`, by the `GetProxiesInfo` method, but that method is new from the multiproxy PR, so we wouldn't be breaking anybody's expectations by removing `StatsHandler` from `ProxyInfo`.

File: cmd/utils/flags.go (hunk 1 above)
Comment (gastonponti) [reply cmd/utils/flags.go:783]: Yes, I thought about that, but my concern is that the flag only lives in the multi-proxy branch. So technically was not released to everyone.
I don't want to add something that will start as deprecated. @kevjue what do you think about this?

File: ethstats/ethstats.go (hunk 3 above)
Comment (gastonponti) [reply ethstats/ethstats.go:574]: This was a change of the #1033. Thought that maybe was important, and I was not seeing the why

File: cmd/utils/flags.go (hunk 1 above)
Comment (trianglesphere) [reply cmd/utils/flags.go:783]: If it was just from multi proxy, I definitely agree that we should remove it.

**Review by mcortesi — APPROVED** [review]

File: ethstats/ethstats.go [hunk 8]
```diff
@@ -188,166 +207,116 @@ func (s *Service) loop() {
...
-				if time.Duration(mclock.Now()-lastTx) < time.Second {
-					continue
+			}
+		})
+	} else {
+		group.Go(func() error {
+			// Loop reporting until termination
+			for {
+				// Resolve the URL, defaulting to TLS, but falling back to none too
+				path := fmt.Sprintf("%s/api", s.celostatsHost)
```
[the remark this answers is not in the document]
Comment (gastonponti) [reply ethstats/ethstats.go]: mmmm It's already out... Maybe an outdated file from github?
